pkg/image/controller/controller.go

```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+					continue
+				}
+				imports[tagName] = ref.DockerClientDefaults()
+			} else {
+				references.Insert(tagName)
+			}
+		}
+	}
+
+	if len(stream.Spec.DockerImageRepository) == 0 {
+		return imports, nil
+	}
+
+	// read tags from default upstream image repository
+	streamRef, err := api.DockerImageReferenceForStream(stream)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := client.Connect(streamRef.Registry, insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	tags, err := conn.ImageTags(streamRef.Namespace, streamRef.Name)
 	switch {
 	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-		return c.done(stream, err.Error(), retryCount)
+		return nil, c.done(stream, err.Error(), retryCount)
 	case err != nil:
-		return err
+		return nil, err
 	}
 
-	imageToTag := make(map[string][]string)
 	for tag, image := range tags {
-		if specTag, ok := stream.Spec.Tags[tag]; ok && specTag.From != nil {
-			// spec tag is set to track another tag - do not import
+		if _, ok := imports[tag]; ok {
+			continue
+		}
+		if references.Has(tag) {
 			continue
 		}
 
-		imageToTag[image] = append(imageToTag[image], tag)
-	}
-
-	// no tags to import
-	if len(imageToTag) == 0 {
-		return c.done(stream, "", retryCount)
+		idTagPresent := false
+		for t2, i2 := range tags {
+			if t2 == tag {
+				continue
+			}
+			if i2 == image && t2 == image {
+				idTagPresent = true
+				break
+			}
+		}
+		ref := streamRef
+		if idTagPresent {
+			ref.Tag = image
+		} else {
+			ref.Tag = tag
+		}
+		ref.ID = image
+		imports[tag] = ref
 	}
+	return imports, nil
+}
 
-	for id, tags := range imageToTag {
-		dockerImage, err := conn.ImageByID(ref.Namespace, ref.Name, id)
-		switch {
-		case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-			return c.done(stream, err.Error(), retryCount)
-		case dockerregistry.IsImageNotFound(err):
-			continue
-		case err != nil:
-			return err
+// importTags imports tags specified in a map from given ImageStream. Returns an error if one occurs.
+func (c *ImportController) importTags(stream *api.ImageStream, imports map[string]api.DockerImageReference, client dockerregistry.Client, insecure bool) error {
+	retrieved := make(map[string]*dockerregistry.Image)
+	for tag, ref := range imports {
+		var dockerImage *dockerregistry.Image
+		if image, ok := retrieved[ref.ID]; ok {
+			dockerImage = image
+		} else {
+			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
+			conn, err := client.Connect(ref.Registry, insecure)
+			if err != nil {
+				return err
+			}
+			if len(ref.ID) > 0 {
+				dockerImage, err = conn.ImageByID(ref.Namespace, ref.Name, ref.ID)
+			} else {
+				dockerImage, err = conn.ImageByTag(ref.Namespace, ref.Name, ref.Tag)
+			}
+			switch {
+			case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
```

Comment:
This is no longer the case where it should terminate an import.  Instead, you need to make as much progress as possible and then fail at the end.  It would be better to accumulate errors and continue.  `c.done()` should not be invoked in this loop.  This iteration of the loop should be extracted into a function.